From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 22 Jan 2025 17:23:16 +0800
Subject: [PATCH] Gale Optimize noise generation

Co-authored by: Martijn Muijsers <martijnmuijsers@live.nl>
ishland <ishlandmc@yeah.net>
As part of: Gale (https://github.com/GaleMC/Gale/blob/276e903b2688f23b19bdc8d493c0bf87656d2400/patches/server/0106-Optimize-noise-generation.patch)
C2ME (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT (https://opensource.org/licenses/MIT)

diff --git a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
index ef8f2a00945a3d240a23ca73f138960ee0fe8742..8627e40502245d4688b989996ce1efdf9151bff2 100644
--- a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -11,6 +11,27 @@ public final class ImprovedNoise {
     public final double yo;
     public final double zo;
 
+    // Gale start - C2ME - optimize noise generation
+    private static final double[] FLAT_SIMPLEX_GRAD = new double[]{
+        1, 1, 0, 0,
+        -1, 1, 0, 0,
+        1, -1, 0, 0,
+        -1, -1, 0, 0,
+        1, 0, 1, 0,
+        -1, 0, 1, 0,
+        1, 0, -1, 0,
+        -1, 0, -1, 0,
+        0, 1, 1, 0,
+        0, -1, 1, 0,
+        0, 1, -1, 0,
+        0, -1, -1, 0,
+        1, 1, 0, 0,
+        0, -1, 1, 0,
+        -1, 1, 0, 0,
+        0, -1, -1, 0,
+    };
+    // Gale end - C2ME - optimize noise generation
+
     public ImprovedNoise(final RandomSource random) {
         this.xo = random.nextDouble() * 256.0;
         this.yo = random.nextDouble() * 256.0;
@@ -38,9 +59,11 @@ public final class ImprovedNoise {
         double x = _x + this.xo;
         double y = _y + this.yo;
         double z = _z + this.zo;
-        int xf = Mth.floor(x);
-        int yf = Mth.floor(y);
-        int zf = Mth.floor(z);
+        // Gale start - C2ME - optimize noise generation - optimize: remove frequent type conversions
+        double xf = Math.floor(x);
+        double yf = Math.floor(y);
+        double zf = Math.floor(z);
+        // Gale end - C2ME - optimize noise generation - optimize: remove frequent type conversions
         double xr = x - xf;
         double yr = y - yf;
         double zr = z - zf;
@@ -53,25 +76,27 @@ public final class ImprovedNoise {
                 fudgeLimit = yr;
             }
 
-            yrFudge = Mth.floor(fudgeLimit / yScale + 1.0E-7F) * yScale;
+            yrFudge = Math.floor(fudgeLimit / yScale + 1.0E-7F) * yScale; // Gale - C2ME - optimize noise generation - optimize: remove frequent type conversions
         } else {
             yrFudge = 0.0;
         }
 
-        return this.sampleAndLerp(xf, yf, zf, xr, yr - yrFudge, zr, yr);
+        return this.sampleAndLerp((int) xf, (int) yf, (int) zf, xr, yr - yrFudge, zr, yr); // Gale - C2ME - optimize noise generation - optimize: remove frequent type conversions
     }
 
     public double noiseWithDerivative(final double _x, final double _y, final double _z, final double[] derivativeOut) {
         double x = _x + this.xo;
         double y = _y + this.yo;
         double z = _z + this.zo;
-        int xf = Mth.floor(x);
-        int yf = Mth.floor(y);
-        int zf = Mth.floor(z);
+        // Gale start - C2ME - optimize noise generation - optimize: remove frequent type conversions
+        double xf = Math.floor(x);
+        double yf = Math.floor(y);
+        double zf = Math.floor(z);
+        // Gale end - C2ME - optimize noise generation - optimize: remove frequent type conversions
         double xr = x - xf;
         double yr = y - yf;
         double zr = z - zf;
-        return this.sampleWithDerivative(xf, yf, zf, xr, yr, zr, derivativeOut);
+        return this.sampleWithDerivative((int) xf, (int) yf, (int) zf, xr, yr, zr, derivativeOut); // Gale - C2ME - optimize noise generation - optimize: remove frequent type conversions
     }
 
     private static double gradDot(final int hash, final double x, final double y, final double z) {
@@ -83,24 +108,69 @@ public final class ImprovedNoise {
     }
 
     private double sampleAndLerp(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double yrOriginal) {
-        int x0 = this.p(x);
-        int x1 = this.p(x + 1);
-        int xy00 = this.p(x0 + y);
-        int xy01 = this.p(x0 + y + 1);
-        int xy10 = this.p(x1 + y);
-        int xy11 = this.p(x1 + y + 1);
-        double d000 = gradDot(this.p(xy00 + z), xr, yr, zr);
-        double d100 = gradDot(this.p(xy10 + z), xr - 1.0, yr, zr);
-        double d010 = gradDot(this.p(xy01 + z), xr, yr - 1.0, zr);
-        double d110 = gradDot(this.p(xy11 + z), xr - 1.0, yr - 1.0, zr);
-        double d001 = gradDot(this.p(xy00 + z + 1), xr, yr, zr - 1.0);
-        double d101 = gradDot(this.p(xy10 + z + 1), xr - 1.0, yr, zr - 1.0);
-        double d011 = gradDot(this.p(xy01 + z + 1), xr, yr - 1.0, zr - 1.0);
-        double d111 = gradDot(this.p(xy11 + z + 1), xr - 1.0, yr - 1.0, zr - 1.0);
-        double xAlpha = Mth.smoothstep(xr);
-        double yAlpha = Mth.smoothstep(yrOriginal);
-        double zAlpha = Mth.smoothstep(zr);
-        return Mth.lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
+        // Gale start - C2ME - optimize noise generation - inline math & small optimization: remove frequent type conversions and redundant ops
+        final int var0 = x & 0xFF;
+        final int var1 = (x + 1) & 0xFF;
+        final int var2 = this.p[var0] & 0xFF;
+        final int var3 = this.p[var1] & 0xFF;
+        final int var4 = (var2 + y) & 0xFF;
+        final int var5 = (var3 + y) & 0xFF;
+        final int var6 = (var2 + y + 1) & 0xFF;
+        final int var7 = (var3 + y + 1) & 0xFF;
+        final int var8 = this.p[var4] & 0xFF;
+        final int var9 = this.p[var5] & 0xFF;
+        final int var10 = this.p[var6] & 0xFF;
+        final int var11 = this.p[var7] & 0xFF;
+
+        final int var12 = (var8 + z) & 0xFF;
+        final int var13 = (var9 + z) & 0xFF;
+        final int var14 = (var10 + z) & 0xFF;
+        final int var15 = (var11 + z) & 0xFF;
+        final int var16 = (var8 + z + 1) & 0xFF;
+        final int var17 = (var9 + z + 1) & 0xFF;
+        final int var18 = (var10 + z + 1) & 0xFF;
+        final int var19 = (var11 + z + 1) & 0xFF;
+        final int var20 = (this.p[var12] & 15) << 2;
+        final int var21 = (this.p[var13] & 15) << 2;
+        final int var22 = (this.p[var14] & 15) << 2;
+        final int var23 = (this.p[var15] & 15) << 2;
+        final int var24 = (this.p[var16] & 15) << 2;
+        final int var25 = (this.p[var17] & 15) << 2;
+        final int var26 = (this.p[var18] & 15) << 2;
+        final int var27 = (this.p[var19] & 15) << 2;
+        final double var60 = xr - 1.0;
+        final double var61 = yr - 1.0;
+        final double var62 = zr - 1.0;
+        final double var87 = FLAT_SIMPLEX_GRAD[(var20) | 0] * xr + FLAT_SIMPLEX_GRAD[(var20) | 1] * yr + FLAT_SIMPLEX_GRAD[(var20) | 2] * zr;
+        final double var88 = FLAT_SIMPLEX_GRAD[(var21) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var21) | 1] * yr + FLAT_SIMPLEX_GRAD[(var21) | 2] * zr;
+        final double var89 = FLAT_SIMPLEX_GRAD[(var22) | 0] * xr + FLAT_SIMPLEX_GRAD[(var22) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var22) | 2] * zr;
+        final double var90 = FLAT_SIMPLEX_GRAD[(var23) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var23) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var23) | 2] * zr;
+        final double var91 = FLAT_SIMPLEX_GRAD[(var24) | 0] * xr + FLAT_SIMPLEX_GRAD[(var24) | 1] * yr + FLAT_SIMPLEX_GRAD[(var24) | 2] * var62;
+        final double var92 = FLAT_SIMPLEX_GRAD[(var25) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var25) | 1] * yr + FLAT_SIMPLEX_GRAD[(var25) | 2] * var62;
+        final double var93 = FLAT_SIMPLEX_GRAD[(var26) | 0] * xr + FLAT_SIMPLEX_GRAD[(var26) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var26) | 2] * var62;
+        final double var94 = FLAT_SIMPLEX_GRAD[(var27) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var27) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var27) | 2] * var62;
+
+        final double var95 = xr * 6.0 - 15.0;
+        final double var96 = yrOriginal * 6.0 - 15.0;
+        final double var97 = zr * 6.0 - 15.0;
+        final double var98 = xr * var95 + 10.0;
+        final double var99 = yrOriginal * var96 + 10.0;
+        final double var100 = zr * var97 + 10.0;
+        final double var101 = xr * xr * xr * var98;
+        final double var102 = yrOriginal * yrOriginal * yrOriginal * var99;
+        final double var103 = zr * zr * zr * var100;
+
+        final double var113 = var87 + var101 * (var88 - var87);
+        final double var114 = var93 + var101 * (var94 - var93);
+        final double var115 = var91 + var101 * (var92 - var91);
+        final double var116 = var89 + var101 * (var90 - var89);
+        final double var117 = var114 - var115;
+        final double var118 = var102 * (var116 - var113);
+        final double var119 = var102 * var117;
+        final double var120 = var113 + var118;
+        final double var121 = var115 + var119;
+        return var120 + (var103 * (var121 - var120));
+        // Gale end - C2ME - optimize noise generation - inline math & small optimization: remove frequent type conversions and redundant ops
     }
 
     private double sampleWithDerivative(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double[] derivativeOut) {
diff --git a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java.rej b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java.rej
new file mode 100644
index 0000000000000000000000000000000000000000..1bb8c8065690005d604fcc4905bed387fe4aa9d7
--- /dev/null
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java.rej
@@ -0,0 +1,166 @@
+diff a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java	(rejected hunks)
+@@ -11,6 +11,27 @@ public final class ImprovedNoise {
+     public final double yo;
+     public final double zo;
+ 
++    // Gale start - C2ME - optimize noise generation
++    private static final double[] FLAT_SIMPLEX_GRAD = new double[]{
++        1, 1, 0, 0,
++        -1, 1, 0, 0,
++        1, -1, 0, 0,
++        -1, -1, 0, 0,
++        1, 0, 1, 0,
++        -1, 0, 1, 0,
++        1, 0, -1, 0,
++        -1, 0, -1, 0,
++        0, 1, 1, 0,
++        0, -1, 1, 0,
++        0, 1, -1, 0,
++        0, -1, -1, 0,
++        1, 1, 0, 0,
++        0, -1, 1, 0,
++        -1, 1, 0, 0,
++        0, -1, -1, 0,
++    };
++    // Gale end - C2ME - optimize noise generation
++
+     public ImprovedNoise(RandomSource random) {
+         this.xo = random.nextDouble() * 256.0;
+         this.yo = random.nextDouble() * 256.0;
+@@ -38,9 +59,11 @@ public final class ImprovedNoise {
+         double d = x + this.xo;
+         double d1 = y + this.yo;
+         double d2 = z + this.zo;
+-        int floor = Mth.floor(d);
+-        int floor1 = Mth.floor(d1);
+-        int floor2 = Mth.floor(d2);
++        // Gale start - C2ME - optimize noise generation - optimize: remove frequent type conversions
++        double floor = Math.floor(d);
++        double floor1 = Math.floor(d1);
++        double floor2 = Math.floor(d2);
++        // Gale end - C2ME - optimize noise generation - optimize: remove frequent type conversions
+         double d3 = d - floor;
+         double d4 = d1 - floor1;
+         double d5 = d2 - floor2;
+@@ -53,25 +76,27 @@ public final class ImprovedNoise {
+                 d6 = d4;
+             }
+ 
+-            d7 = Mth.floor(d6 / yScale + 1.0E-7F) * yScale;
++            d7 = Math.floor(d6 / yScale + 1.0E-7F) * yScale; // Gale - C2ME - optimize noise generation - optimize: remove frequent type conversions
+         } else {
+             d7 = 0.0;
+         }
+ 
+-        return this.sampleAndLerp(floor, floor1, floor2, d3, d4 - d7, d5, d4);
++        return this.sampleAndLerp((int) floor, (int) floor1, (int) floor2, d3, d4 - d7, d5, d4); // Gale - C2ME - optimize noise generation - optimize: remove frequent type conversions
+     }
+ 
+     public double noiseWithDerivative(double x, double y, double z, double[] values) {
+         double d = x + this.xo;
+         double d1 = y + this.yo;
+         double d2 = z + this.zo;
+-        int floor = Mth.floor(d);
+-        int floor1 = Mth.floor(d1);
+-        int floor2 = Mth.floor(d2);
++        // Gale start - C2ME - optimize noise generation - optimize: remove frequent type conversions
++        double floor = Math.floor(d);
++        double floor1 = Math.floor(d1);
++        double floor2 = Math.floor(d2);
++        // Gale end - C2ME - optimize noise generation - optimize: remove frequent type conversions
+         double d3 = d - floor;
+         double d4 = d1 - floor1;
+         double d5 = d2 - floor2;
+-        return this.sampleWithDerivative(floor, floor1, floor2, d3, d4, d5, values);
++        return this.sampleWithDerivative((int) floor, (int) floor1, (int) floor2, d3, d4, d5, values); // Gale - C2ME - optimize noise generation - optimize: remove frequent type conversions
+     }
+ 
+     private static double gradDot(int gradIndex, double xFactor, double yFactor, double zFactor) {
+@@ -83,24 +108,69 @@ public final class ImprovedNoise {
+     }
+ 
+     private double sampleAndLerp(int gridX, int gridY, int gridZ, double deltaX, double weirdDeltaY, double deltaZ, double deltaY) {
+-        int i = this.p(gridX);
+-        int i1 = this.p(gridX + 1);
+-        int i2 = this.p(i + gridY);
+-        int i3 = this.p(i + gridY + 1);
+-        int i4 = this.p(i1 + gridY);
+-        int i5 = this.p(i1 + gridY + 1);
+-        double d = gradDot(this.p(i2 + gridZ), deltaX, weirdDeltaY, deltaZ);
+-        double d1 = gradDot(this.p(i4 + gridZ), deltaX - 1.0, weirdDeltaY, deltaZ);
+-        double d2 = gradDot(this.p(i3 + gridZ), deltaX, weirdDeltaY - 1.0, deltaZ);
+-        double d3 = gradDot(this.p(i5 + gridZ), deltaX - 1.0, weirdDeltaY - 1.0, deltaZ);
+-        double d4 = gradDot(this.p(i2 + gridZ + 1), deltaX, weirdDeltaY, deltaZ - 1.0);
+-        double d5 = gradDot(this.p(i4 + gridZ + 1), deltaX - 1.0, weirdDeltaY, deltaZ - 1.0);
+-        double d6 = gradDot(this.p(i3 + gridZ + 1), deltaX, weirdDeltaY - 1.0, deltaZ - 1.0);
+-        double d7 = gradDot(this.p(i5 + gridZ + 1), deltaX - 1.0, weirdDeltaY - 1.0, deltaZ - 1.0);
+-        double d8 = Mth.smoothstep(deltaX);
+-        double d9 = Mth.smoothstep(deltaY);
+-        double d10 = Mth.smoothstep(deltaZ);
+-        return Mth.lerp3(d8, d9, d10, d, d1, d2, d3, d4, d5, d6, d7);
++        // Gale start - C2ME - optimize noise generation - inline math & small optimization: remove frequent type conversions and redundant ops
++        final int var0 = gridX & 0xFF;
++        final int var1 = (gridX + 1) & 0xFF;
++        final int var2 = this.p[var0] & 0xFF;
++        final int var3 = this.p[var1] & 0xFF;
++        final int var4 = (var2 + gridY) & 0xFF;
++        final int var5 = (var3 + gridY) & 0xFF;
++        final int var6 = (var2 + gridY + 1) & 0xFF;
++        final int var7 = (var3 + gridY + 1) & 0xFF;
++        final int var8 = this.p[var4] & 0xFF;
++        final int var9 = this.p[var5] & 0xFF;
++        final int var10 = this.p[var6] & 0xFF;
++        final int var11 = this.p[var7] & 0xFF;
++
++        final int var12 = (var8 + gridZ) & 0xFF;
++        final int var13 = (var9 + gridZ) & 0xFF;
++        final int var14 = (var10 + gridZ) & 0xFF;
++        final int var15 = (var11 + gridZ) & 0xFF;
++        final int var16 = (var8 + gridZ + 1) & 0xFF;
++        final int var17 = (var9 + gridZ + 1) & 0xFF;
++        final int var18 = (var10 + gridZ + 1) & 0xFF;
++        final int var19 = (var11 + gridZ + 1) & 0xFF;
++        final int var20 = (this.p[var12] & 15) << 2;
++        final int var21 = (this.p[var13] & 15) << 2;
++        final int var22 = (this.p[var14] & 15) << 2;
++        final int var23 = (this.p[var15] & 15) << 2;
++        final int var24 = (this.p[var16] & 15) << 2;
++        final int var25 = (this.p[var17] & 15) << 2;
++        final int var26 = (this.p[var18] & 15) << 2;
++        final int var27 = (this.p[var19] & 15) << 2;
++        final double var60 = deltaX - 1.0;
++        final double var61 = weirdDeltaY - 1.0;
++        final double var62 = deltaZ - 1.0;
++        final double var87 = FLAT_SIMPLEX_GRAD[(var20) | 0] * deltaX + FLAT_SIMPLEX_GRAD[(var20) | 1] * weirdDeltaY + FLAT_SIMPLEX_GRAD[(var20) | 2] * deltaZ;
++        final double var88 = FLAT_SIMPLEX_GRAD[(var21) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var21) | 1] * weirdDeltaY + FLAT_SIMPLEX_GRAD[(var21) | 2] * deltaZ;
++        final double var89 = FLAT_SIMPLEX_GRAD[(var22) | 0] * deltaX + FLAT_SIMPLEX_GRAD[(var22) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var22) | 2] * deltaZ;
++        final double var90 = FLAT_SIMPLEX_GRAD[(var23) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var23) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var23) | 2] * deltaZ;
++        final double var91 = FLAT_SIMPLEX_GRAD[(var24) | 0] * deltaX + FLAT_SIMPLEX_GRAD[(var24) | 1] * weirdDeltaY + FLAT_SIMPLEX_GRAD[(var24) | 2] * var62;
++        final double var92 = FLAT_SIMPLEX_GRAD[(var25) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var25) | 1] * weirdDeltaY + FLAT_SIMPLEX_GRAD[(var25) | 2] * var62;
++        final double var93 = FLAT_SIMPLEX_GRAD[(var26) | 0] * deltaX + FLAT_SIMPLEX_GRAD[(var26) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var26) | 2] * var62;
++        final double var94 = FLAT_SIMPLEX_GRAD[(var27) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var27) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var27) | 2] * var62;
++
++        final double var95 = deltaX * 6.0 - 15.0;
++        final double var96 = deltaY * 6.0 - 15.0;
++        final double var97 = deltaZ * 6.0 - 15.0;
++        final double var98 = deltaX * var95 + 10.0;
++        final double var99 = deltaY * var96 + 10.0;
++        final double var100 = deltaZ * var97 + 10.0;
++        final double var101 = deltaX * deltaX * deltaX * var98;
++        final double var102 = deltaY * deltaY * deltaY * var99;
++        final double var103 = deltaZ * deltaZ * deltaZ * var100;
++
++        final double var113 = var87 + var101 * (var88 - var87);
++        final double var114 = var93 + var101 * (var94 - var93);
++        final double var115 = var91 + var101 * (var92 - var91);
++        final double var116 = var89 + var101 * (var90 - var89);
++        final double var117 = var114 - var115;
++        final double var118 = var102 * (var116 - var113);
++        final double var119 = var102 * var117;
++        final double var120 = var113 + var118;
++        final double var121 = var115 + var119;
++        return var120 + (var103 * (var121 - var120));
++        // Gale end - C2ME - optimize noise generation - inline math & small optimization: remove frequent type conversions and redundant ops
+     }
+ 
+     private double sampleWithDerivative(int gridX, int gridY, int gridZ, double deltaX, double deltaY, double deltaZ, double[] noiseValues) {
diff --git a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
index dc632666201028e680443eb27f235583506349ad..e177f6783579e39afbbacec0f5f0961d4fa0bacf 100644
--- a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
@@ -26,6 +26,10 @@ public class PerlinNoise {
     private final double lowestFreqValueFactor;
     private final double lowestFreqInputFactor;
     private final double maxValue;
+    // Gale start - C2ME - optimize noise generation
+    private final int octaveSamplersCount;
+    private final double [] amplitudesArray;
+    // Gale end - C2ME - optimize noise generation
 
     @Deprecated
     public static PerlinNoise createLegacyForBlendedNoise(final RandomSource random, final IntStream octaves) {
@@ -127,6 +131,10 @@ public class PerlinNoise {
         this.lowestFreqInputFactor = Math.pow(2.0, -zeroOctaveIndex);
         this.lowestFreqValueFactor = Math.pow(2.0, octaves - 1) / (Math.pow(2.0, octaves) - 1.0);
         this.maxValue = this.edgeValue(2.0);
+        // Gale start - C2ME - optimize noise generation
+        this.octaveSamplersCount = this.noiseLevels.length;
+        this.amplitudesArray = this.amplitudes.toDoubleArray();
+        // Gale end - C2ME - optimize noise generation
     }
 
     protected double maxValue() {
@@ -138,7 +146,27 @@ public class PerlinNoise {
     }
 
     public double getValue(final double x, final double y, final double z) {
-        return this.getValue(x, y, z, 0.0, 0.0);
+        // Gale start - C2ME - optimize noise generation - optimize for common cases
+        double d = 0.0;
+        double e = this.lowestFreqInputFactor;
+        double f = this.lowestFreqValueFactor;
+
+        for (int i = 0; i < this.octaveSamplersCount; ++i) {
+            ImprovedNoise perlinNoiseSampler = this.noiseLevels[i];
+            if (perlinNoiseSampler != null) {
+                @SuppressWarnings("deprecation")
+                double g = perlinNoiseSampler.noise(
+                    wrap(x * e), wrap(y * e), wrap(z * e), 0.0, 0.0
+                );
+                d += this.amplitudesArray[i] * g * f;
+            }
+
+            e *= 2.0;
+            f /= 2.0;
+        }
+
+        return d;
+        // Gale end - C2ME - optimize noise generation - optimize for common cases
     }
 
     @Deprecated
diff --git a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java.rej b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java.rej
new file mode 100644
index 0000000000000000000000000000000000000000..12d717fda05704641c037af09667429d37072dec
--- /dev/null
+++ b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java.rej
@@ -0,0 +1,52 @@
+diff a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java	(rejected hunks)
+@@ -26,6 +26,10 @@ public class PerlinNoise {
+     private final double lowestFreqValueFactor;
+     private final double lowestFreqInputFactor;
+     private final double maxValue;
++    // Gale start - C2ME - optimize noise generation
++    private final int octaveSamplersCount;
++    private final double [] amplitudesArray;
++    // Gale end - C2ME - optimize noise generation
+ 
+     @Deprecated
+     public static PerlinNoise createLegacyForBlendedNoise(RandomSource random, IntStream octaves) {
+@@ -127,6 +131,10 @@ public class PerlinNoise {
+         this.lowestFreqInputFactor = Math.pow(2.0, -i);
+         this.lowestFreqValueFactor = Math.pow(2.0, size - 1) / (Math.pow(2.0, size) - 1.0);
+         this.maxValue = this.edgeValue(2.0);
++        // Gale start - C2ME - optimize noise generation
++        this.octaveSamplersCount = this.noiseLevels.length;
++        this.amplitudesArray = this.amplitudes.toDoubleArray();
++        // Gale end - C2ME - optimize noise generation
+     }
+ 
+     protected double maxValue() {
+@@ -138,7 +146,27 @@ public class PerlinNoise {
+     }
+ 
+     public double getValue(double x, double y, double z) {
+-        return this.getValue(x, y, z, 0.0, 0.0, false);
++        // Gale start - C2ME - optimize noise generation - optimize for common cases
++        double d = 0.0;
++        double e = this.lowestFreqInputFactor;
++        double f = this.lowestFreqValueFactor;
++
++        for (int i = 0; i < this.octaveSamplersCount; ++i) {
++            ImprovedNoise perlinNoiseSampler = this.noiseLevels[i];
++            if (perlinNoiseSampler != null) {
++                @SuppressWarnings("deprecation")
++                double g = perlinNoiseSampler.noise(
++                    wrap(x * e), wrap(y * e), wrap(z * e), 0.0, 0.0
++                );
++                d += this.amplitudesArray[i] * g * f;
++            }
++
++            e *= 2.0;
++            f /= 2.0;
++        }
++
++        return d;
++        // Gale end - C2ME - optimize noise generation - optimize for common cases
+     }
+ 
+     @Deprecated
